### PR TITLE
Type members

### PR DESCRIPTION
**Pull request by RustanLeino**
This is a major feature addition. It allows `datatype`, `codatatype`, and `newtype` type declarations to declare members (like `function`, `method`, and `const`) by following the type-signature declaration with a curly-brace body as for a `class`.

This pull request consists of four commits, for syntax, type checking, verification, and compilation, respectively.

### CODE REVIEW

**Comment by seanmcl** [comment]
What is the motivation for this change?  This adds complexity to the language.  Is it just so we can have dot notation, rather than having functions that operate on the datatypes a la ML, or is there something we can do now we couldn't do before?

File: Test/cloudmake/CloudMake-ConsistentBuilds.dfy [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %dafny /compile:0 /dprint:"%t.dprint" /autoTriggers:0 "%s" > "%t"
+// RUN: %dafny /compile:0 /dprint:"%t.dprint" /autoTriggers:0 /restartProver "%s" > "%t"
```
Comment (seanmcl) [inline Test/cloudmake/CloudMake-ConsistentBuilds.dfy]: Is this related to type members?

File: Source/Dafny/Compiler-Csharp.cs [hunk 2]
```diff
@@ -705,17 +720,28 @@ protected class ClassWriter : IClassWriter {
         }
       }
 
+      var customReceiver = !m.IsStatic && m.EnclosingClass is NewtypeDecl;
```
Comment (seanmcl) [inline Source/Dafny/Compiler-Csharp.cs]: This repeats 5 times in the review.  Could we add it to `Method` perhaps?

File: Source/Dafny/Compiler-go.cs [hunk 3]
```diff
@@ -1135,44 +1146,6 @@ protected class ClassWriter : IClassWriter {
       }
     }
 
-    List<TypeParameter> UsedTypeParameters(DatatypeDecl dt) {
```
Comment (seanmcl) [inline Source/Dafny/Compiler-go.cs]: Code movement changes together with big logical changes make code review difficult.

File: Test/dafny0/NameclashesCompile.dfy [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  print c.c, " ", c.d, " ", c.Computer(), " ";
+  print r0, " ", r1, "\n";
+  
+  var n: NameclashNew := 3;
+  r0 := n.ToString();
+  r1 := n.GetHashCode();
+  print r0, " ", r1, "\n";
+}
+
+datatype Nameclash = Ctor(x: int)
```
Comment (seanmcl) [inline Test/dafny0/NameclashesCompile.dfy]: I don't understand the example.  Where is the name clash?

File: Test/dafny0/TypeMembers.dfy [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  BasicTests();
+  MoreTests();
+}
+
+// ---------- basic tests ----------
+
+method BasicTests() {
+  var t := Yes;
+  var r := t.W();  // 10
+  var q := t.Q() + DaTy.Q();  // 13 + 13
```
Comment (seanmcl) [inline Test/dafny0/TypeMembers.dfy]: I'm surprised `t.Q()` works here.  I tried to find this in the reference manual but failed.  3.5.0 was the only relevant bit I could find.  Is it in there and I missed it, or should we update the manual?

**Comment by samuelgruetter** [comment]
> What is the motivation for this change?  This adds complexity to the language.  Is it just so we can have dot notation, rather than having functions that operate on the datatypes a la ML, or is there something we can do now we couldn't do before?

My first reaction was exactly the same! However, after some more reflection, here's why I believe that this feature is worth adding:
With current Dafny, if I have three datatypes `A`, `B`, and `C`, and I want to define a method `foo` for each of them, I'd have to call them `foo_A`, `foo_B`, and `foo_C`. Now I could ask for a new feature like overloaded methods, so that I can define three methods called `foo`, and static dispatch decides which of them to call. But it turns out that this PR solves this in a much nicer and more understandable way: The type of the expression on the left-hand side of the dot determines which "overload" to call.

So the "thing we couldn't do before" is having one method called `foo` for three different datatypes. 

And for another piece of motivation, adding exceptions will need this, see https://github.com/dafny-lang/rfcs/blob/250686552d3c2aeec8fc471229083a9f1e2c2047/Exceptions.md The methods `IsFailure`, `PropagateFailure`, and `Extract` have the same problem as `foo` in my above example.

**Comment by RustanLeino** [comment]
The feature gives a way to associate methods/functions with a particular type. As @samuelgruetter said, this gives a nice way to look up certain properties of a type, such as an `IsFailure` member. It also paves the way for a future extension of `trait`, to allow `trait` to describe not just reference types but any types.

File: Test/cloudmake/CloudMake-ConsistentBuilds.dfy (hunk 1 above)
Comment (RustanLeino) [reply Test/cloudmake/CloudMake-ConsistentBuilds.dfy]: No, only tangentially. See the commit comment for this change.

File: Test/dafny0/NameclashesCompile.dfy (hunk 4 above)
Comment (RustanLeino) [reply Test/dafny0/NameclashesCompile.dfy]: There are two sets of name clashes. One is the names `ToString` and `GetHashCode`, which exist in .NET as overridable methods for all classes. The Dafny compiler for C# previously passed these names through, which causes the C# compiler to complain that there's a method that shadows a virtual method (which would commonly be a mistake). The other set of name clashes comes from the fields `c` and `d` that the Dafny-to-C# compiler generates for `codatatype`s.

This file thus contains regression tests for these previous bugs.

File: Source/Dafny/Compiler-Csharp.cs (hunk 2 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler-Csharp.cs]: Good idea.

File: Test/dafny0/TypeMembers.dfy (hunk 5 above)
Comment (RustanLeino) [reply Test/dafny0/TypeMembers.dfy]: Dafny allows static members of a class to be referenced by an expression whose type is the class. Other than being used to look up the static member, the given receiver expression is completed ignored (so it may even be an expression that wouldn't be well-formed in other contexts).

It would be good to update the manual on this (and many other) point(s).

**Comment by RustanLeino** [comment]
The previous force-pushed was a re-base with the new, big java-compiler commit.
